module/mempool/stdmap/approvals.go

```diff
@@ -195,36 +173,65 @@ func (a *Approvals) ByChunk(resultID flow.Identifier, chunkIndex uint64) map[flo
 	// determine the lookup key for the corresponding chunk
 	chunkKey := key(resultID, chunkIndex)
 
-	entity, exists := a.backend.ByID(chunkKey)
-	if !exists {
-		return nil
-	}
-
-	approvalMapEntity, ok := entity.(model.ApprovalMapEntity)
-	if !ok {
+	// To guarantee concurrency safety, we need to copy the map via a locked operation in the backend.
+	// Otherwise, another routine might concurrently modify the map stored for the same resultID.
+	approvals := make(map[flow.Identifier]*flow.ResultApproval)
+	err := a.backend.Run(func(backdata map[flow.Identifier]flow.Entity) error {
+		entity, exists := backdata[chunkKey]
+		if !exists {
+			return nil
+		}
+		approvalMapEntity, ok := entity.(model.ApprovalMapEntity)
+		if !ok {
+			return fmt.Errorf("unexpected entity type %T", entity)
+		}
+		for i, app := range approvalMapEntity.Approvals {
+			approvals[i] = app
+		}
 		return nil
+	})
+	if err != nil {
+		// The current implementation never reaches this path, as it only stores
+		// ApprovalMapEntity as entities in the mempool. Reaching this error
+		// condition implies this code was inconsistently modified.
+		panic("unexpected internal error in IncorporatedResults mempool: " + err.Error())
 	}
 
-	return approvalMapEntity.Approvals
+	return approvals
 }
 
 // All will return all approvals in the memory pool.
 func (a *Approvals) All() []*flow.ResultApproval {
 	res := make([]*flow.ResultApproval, 0)
 
-	entities := a.backend.All()
-	for _, entity := range entities {
-		approvalMapEntity, _ := entity.(model.ApprovalMapEntity)
-
-		for _, approval := range approvalMapEntity.Approvals {
-			res = append(res, approval)
+	err := a.backend.Run(func(backdata map[flow.Identifier]flow.Entity) error {
+		for _, entity := range backdata {
+			approvalMapEntity, ok := entity.(model.ApprovalMapEntity)
+			if !ok {
+				// should never happen: as the mempool only stores ApprovalMapEntity
+				return fmt.Errorf("unexpected entity type %T", entity)
+			}
```

Comment:
IMO it's not worth it to catch a failed type assertion just to panic, since an _uncaught_ type assertion will panic just fine on its own 😉

Reply:
thanks for the suggestion. Decluttered `stdmap.Approvals` and `stdmap.IncorporatedResults`. Looks much cleaner now 🧹  😃